### PR TITLE
LD prune optimization & question

### PR DESCRIPTION
**Pull request by jbloom22**
@maccum without checks on the requirements, it could be that `mt` is not partitioned by locus in which case LocalLDPrune wouldn't do much and this would be a bug because the `correlation_matrix` and `locally_pruned_rows` would not align:

```
    block_matrix = BlockMatrix.from_entry_expr(normalized_mean_imputed_genotype_expr, block_size=1024)
    correlation_matrix = (block_matrix @ block_matrix.T)

    locally_pruned_rows = locally_pruned_ds.rows()
    locally_pruned_rows = locally_pruned_rows.key_by(contig=locally_pruned_rows.locus.contig)
    locally_pruned_rows = locally_pruned_rows.select(pos=locally_pruned_rows.locus.position)
    entries = correlation_matrix._filtered_entries_table(locally_pruned_rows, bp_window_size, False)
```

I'm still worried about the following. By changing the key (and partition_key) to `contig=locally_pruned_rows.locus.contig`, the execute code for keyBy below will repartition and resort. I believe the resorting won't actually reorder the key equals the partition key, so each resulting partition is de facto sorted. But what guarantee do we have that groups or rows from different original partitions overlapping the contig aren't re-ordered?

```
  def execute(hc: HailContext): TableValue = {
    val tv = child.execute(hc)
    val rvd = if (sort) {
      def resort: OrderedRVD = {
        val orvdType = new OrderedRVDType(keys.take(nPartitionKeys), keys, typ.rowType)
        OrderedRVD.coerce(orvdType, tv.rvd, None, None)
      }

      tv.rvd match {
        case ordered: OrderedRVD =>
          if (ordered.typ.key.startsWith(keys)
            && ordered.typ.partitionKey.length == nPartitionKeys)
            ordered.copy(typ = ordered.typ.copy(key = keys))
          else resort
        case _: UnpartitionedRVD =>
          resort
      }
...
```

Then in UpperIndexBounds, we do `tbl.groupByKey("positions")`. Our implementation of groupByKey should be fine w.r.t. preserving key (group / contig) order. But later:
```
    val relativeUpperIndexBounds = groupPositionsByKey(tbl).map(positions => {
      scala.util.Sorting.quickSort(positions)
      computeUpperIndexBounds(positions, radius)
    })
```
If these positions aren't already sorted, wouldn't changing their order screw up the alignment with the block matrix? (I think they happen to be sorted because we re-keyed from locus to contig to get here).

I think if we verify that the positions are sorted, rather than sorting the positions, then we guarantee the right behavior of this function with respect to not re-ordering rows (at least up to switching rows with the same position, but that doesn't matter for determining the range bounds). What do you think?

### CODE REVIEW

**Comment by jbloom22** [comment]
I changed the block_size to default (4096) because a 1024 means 4x the number of open files per task for `BlockMatrix.write_from_entry_expr`. On UKBB, that's about 500 open files, each with its own write buffer (64MB on by default on GCP) so Hadoop fails spectacularly. Liam reports that `ld_prune` succeeds with 50k samples (and ~280k markers), and succeeds but with lots of failures with 75k samples, and fails completely with 100k samples. He just tried again with 459k samples reducing the write buffer size to 1MB, and it still failed there, but I'm hopeful the larger block size will help. We'll try 100k again soon.
`  --properties 'core:fs.gs.io.buffersize.write=1048576,core:fs.gs.io.buffersize=8388608' \`

**Comment by maccum** [comment]
In `relativeUpperIndexBounds`, the positions won't be sorted unless we sort them, so that cannot be an assertion. Yes, re-keying by contig means we don't have a guarantee that the positions aren't re-ordered within their groups, but since we sort inside `relativeUpperIndexBounds`, this ensures that they are sorted and that they therefore match the ordering in the block matrix. The sorting doesn't screw up the alignment with the block matrix, because the sorting is what guarantees the alignment with the block matrix.

**Comment by jbloom22** [comment]
@maccum I made block_size a parameter, can you take a quick look before i merge? Using the larger block size fixed hadoop failure in UKBB case, but a smaller block size may still be preferable for fewer samples to increase write parallelism, so best to make it accessible.

File: python/hail/methods/statgen.py [hunk 1]
```diff
@@ -2942,23 +2943,36 @@ def ld_prune(call_expr, r2=0.2, bp_window_size=1000000, memory_per_core=256):
         Window size in base pairs (inclusive upper bound).
     memory_per_core : :obj:`int`
         Memory in MB per core for local pruning queue.
+    block_size: :obj:`int`, optional
+        Block size for block matrices in the global stage. Default given by
+        :meth:`.BlockMatrix.default_block_size`. When writing the block matrix
+        of locally-pruned calls, the number of tasks (parallelism) is the number
+        of block rows (``n_variants / block_size``) and the number of open files
+        (blocks written) per task is the number of block columns
+        (``n_samples / block_size``).
```
Comment (maccum) [inline python/hail/methods/statgen.py]: ```Block size of block matrices used in the global pruning stage. Default given by :meth:`.BlockMatrix.default_block_size`. Block size determines the level of parallelism when writing a block matrix of locally-pruned calls. The number of tasks running in parallel will be equal to ``n_variants / block_size`` and the number of open files per task will be equal to ``n_samples / block_size``.```

**Comment by maccum** [comment]
This looks fine. Could possibly put something in the Notes section of the docs to address changing block size if folks are running into errors with Hadoop, but if most people will be successful with the default, then I guess it's not necessary.

**Comment by jbloom22** [comment]
I've reworked the docs, breaking the description into three stages and describing strategies for dealing with the Hadoop write error.

**Pull request merged**